types/report-renderer.d.ts

```diff
@@ -0,0 +1,60 @@
+/**
+ * @license Copyright 2021 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+
+declare global {
+  module LH.Renderer {
+    export function renderFullReport(lhr: LH.Result, options?: ReportRendererOptions): HTMLElement;
+    export function renderReportComponents(
+      lhr: LH.Result,
+      options?: ReportRendererOptions
+    ): ReportComponents;
+
+    export function getFinalScreenshot(): void | string;
+
+    // Render gauge for a category
+    // category handles a bunch of plugin, n/a, and error cases. groupDefinitions only needed for PWA
+    export function renderGaugeForCategory(
+      category: LH.ReportCategory,
+      groupDefinitions?: Record<string, LH.Result.ReportGroup>
+    ): HTMLElement;
+
+    // Extra convience if you have just a category score.
+    export function renderGaugeForScore(num0to1: number): HTMLElement;
+  }
+
+  // TODO(paulirish): Refactor DOM to match. https://github.com/GoogleChrome/lighthouse/issues/12254#issuecomment-877520562
+  type ReportComponents = {
+    topbarEl: HTMLElement;
+    mainEl: HTMLElement;
+    headerEl: HTMLElement;
+    categoriesEl: HTMLElement;
+    footerEl: HTMLElement;
```

Comment:
ah yea these are all components as found in  #12803 . can we use just the component name for the key here?

```
{
    topbar: HTMLElement;
    main: HTMLElement;
    header: HTMLElement;
    category: HTMLElement[];
    footer: HTMLElement;
}
```

except `lh-category` is made via JS and isn't one of our "components" according to #12803 ...

instead of any of this, using a different name here would work for me. `ReportPartials` ? and then rename `renderReportComponents` to `renderPartialReport` or `renderReportParts`

Reply:
stepping back a bit... why would a client not just be able to call `renderFullReport` and pluck out what they need?

Reply:
`renderReportPartials` SGTM

> why would a client not just be able to call renderFullReport and pluck out what they need?

Because the contract is that the element that is returned is a live, LH-owned element that will update itself and if you break it up, all bets are off. If don't want to take over ownership of *all* interactivity in the report, but need to rearrange some layout, `renderReportComponents` allows you to place individual live elements wherever you wish. 

Given our current embedding system, I think the difference is relatively minor, but that won't be true for flow reports.

Reply:
+1 on partials. great call. these are not 1:1 with our new "components" concept so i love using a different term.

Reply:
> makes me question how much these components can actually be live and update themselves


> coordinating scrolling in the categories with the indicator on the sticky header

Is this a technical feasibility question? Or rhetorical/an objection to an API that returns separate elements that share some global state?

> we have these individual components, but the screenshot lightbox still goes into overlayParentEl passed in the options and clicking on screenshots opens it

Is this a request for `screenshotOverlay` as a partial instead? If so, I like that :)

Reply:
> > makes me question how much these components can actually be live and update themselves
> 
> > coordinating scrolling in the categories with the indicator on the sticky header
> 
> Is this a technical feasibility question? Or rhetorical/an objection to an API that returns separate elements that share some global state?

Feasibility and/or asking for specifics :) How are we handling inter-component interactions with this API?

> Is this a request for `screenshotOverlay` as a partial instead? If so, I like that :)

That might make the most sense, yeah. But again with how are the inter-component interactions going to be/should be set up?

Reply:
The specifics vary depending on what deps we can use, but if we're sticking with imperative onlistener changes approach for example... 

```js
function renderHeader(containerEl, rootState, setState, onRender) {
  containerEl.innerHTML = `<h1>Header y:<span>${rootState.y}</span></h1>`
  onRender(rootState => containerEl.querySelector('span').textContent = rootState.x)
  return containerEl
}

function renderCategories(containerEl, rootState, setState, onRender) {
  containerEl.innerHTML = `...`
  containerEl.onscroll = () => setState({y: containerEl.scrollTop})
  return containerEl
}

function renderPartials() {
  const rootState = {y: 0}
  const headerEl = document.createElement('div')
  const categoriesEl = document.createElement('div')
  
  const listeners = []
  const onRender = listener => listeners.push(listener)
  const setState = state => {
    for (const listener of listeners) listener(state)
  }

  return {
    header: renderHeader(headerEl, rootState, setState, onRender),
    categories: renderCategories(categoriesEl, rootState, setState, onRender),
  }
}
```

Reply:
Right, so _is_ the intention that `renderReportPartials()` would return elements bound to each other to replicate the interactions present in what's generated by `renderReport()`? That's what I'm asking. If so, can we write that down as part of this API? :)

This does increase testing surface area, e.g. if the overlay becomes a partial, with the current renderer the screenshot lightbox is [removed on click](https://github.com/GoogleChrome/lighthouse/blob/fe91d5691e6ecbeb8ae1e35c0bc9783443217252/report/renderer/element-screenshot-renderer.js#L193). If someone doesn't use the overlay, every click on a thumbnail would just [keep appending](https://github.com/GoogleChrome/lighthouse/blob/fe91d5691e6ecbeb8ae1e35c0bc9783443217252/report/renderer/element-screenshot-renderer.js#L162) new lightboxes to the detached overlay element.

Obviously there can be some DIY/use at your own risk aspect to this, but the more automatic bindings there are, the greater the chance of this type of thing, and the more combinatoric testing we need to do (vs a system of user-controlled hooks between the components where we can get away with more unit testing).